src/System.Private.ServiceModel/src/System/ServiceModel/Security/SecurityProtocol.cs

```diff
@@ -30,41 +27,619 @@ namespace System.ServiceModel.Security
+                    SecurityTokenParameters tokenParameters = supportingTokens[i].SecurityTokenParameters;
+                    switch (supportingTokens[i].SecurityTokenAttachmentMode)
+                    {
+                        case SecurityTokenAttachmentMode.Signed:
+                            securityHeader.AddSignedSupportingToken(token, tokenParameters);
+                            break;
+                        case SecurityTokenAttachmentMode.Endorsing:
+                            securityHeader.AddEndorsingSupportingToken(token, tokenParameters);
+                            break;
+                        case SecurityTokenAttachmentMode.SignedEncrypted:
+                            securityHeader.AddBasicSupportingToken(token, tokenParameters);
+                            break;
+                        case SecurityTokenAttachmentMode.SignedEndorsing:
+                            securityHeader.AddSignedEndorsingSupportingToken(token, tokenParameters);
+                            break;
+                        default:
+                            Fx.Assert("Unknown token attachment mode " + supportingTokens[i].SecurityTokenAttachmentMode.ToString());
+                            throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new NotSupportedException(SR.Format(SR.UnknownTokenAttachmentMode, supportingTokens[i].SecurityTokenAttachmentMode.ToString())));
+                    }
+                }
+            }
+        }
+
+        internal static async Task<SecurityToken> GetTokenAsync(SecurityTokenProvider provider, EndpointAddress target, TimeSpan timeout)
+        {
+            if (provider == null)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new MessageSecurityException(SR.Format(SR.TokenProviderCannotGetTokensForTarget, target)));
+            }
+
+            SecurityToken token = null;
+
+            try
+            {
+                token = await provider.GetTokenAsync(timeout);
+            }
+            catch (SecurityTokenException exception)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new MessageSecurityException(SR.Format(SR.TokenProviderCannotGetTokensForTarget, target), exception));
+            }
+            catch (SecurityNegotiationException sne)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new SecurityNegotiationException(SR.Format(SR.TokenProviderCannotGetTokensForTarget, target), sne));
+            }
+
+            return token;
+        }
+
+        public abstract Task<Message> SecureOutgoingMessageAsync(Message message, TimeSpan timeout);
+
+        // subclasses that offer correlation should override this version
+        public virtual async Task<(SecurityProtocolCorrelationState, Message)> SecureOutgoingMessageAsync(Message message, TimeSpan timeout, SecurityProtocolCorrelationState correlationState)
```

Comment:
I know that the type is internal, so this isn't really public, but FWIW we've come to rough conclusions for Framework Design Guidelines regarding Tuples.

* Prefer rich/named types over tuples
* Don't write the word (System.)Tuple
* Don't write the word (System.)ValueTuple
* If you feel a tuple is warranted, use named tuples with PascalCase identifiers.